packages/editor/src/lib/components/CulledShapes.tsx

```diff
@@ -0,0 +1,178 @@
+	context.attachShader(program, fragmentShader)
+	context.linkProgram(program)
+	if (!context.getProgramParameter(program, context.LINK_STATUS)) {
+		return
+	}
+	context.useProgram(program)
+
+	const shapeVertexPositionAttributeLocation = context.getAttribLocation(
+		program,
+		'shapeVertexPosition'
+	)
+	if (shapeVertexPositionAttributeLocation < 0) {
+		return
+	}
+	context.enableVertexAttribArray(shapeVertexPositionAttributeLocation)
+
+	const viewportStartUniformLocation = context.getUniformLocation(program, 'viewportStart')
+	const viewportEndUniformLocation = context.getUniformLocation(program, 'viewportEnd')
+	if (!viewportStartUniformLocation || !viewportEndUniformLocation) {
+		return
+	}
+	return {
+		context,
+		program,
+		shapeVertexPositionAttributeLocation,
+		viewportStartUniformLocation,
+		viewportEndUniformLocation,
+	}
+}
+
+export function CulledShapes() {
+	const editor = useEditor()
+	const isDarkMode = useIsDarkMode()
+	const canvasRef = useRef<HTMLCanvasElement>(null)
+
+	const isCullingOffScreenShapes = Number.isFinite(editor.renderingBoundsMargin)
+
+	useEffect(() => {
+		const webGl = setupWebGl(canvasRef.current, isDarkMode)
+		if (!webGl) return
+		if (!isCullingOffScreenShapes) return
+
+		const {
+			context,
+			shapeVertexPositionAttributeLocation,
+			viewportStartUniformLocation,
+			viewportEndUniformLocation,
+		} = webGl
+
+		const shapeVertices = computed('shape vertices', function calculateCulledShapeVertices() {
+			const results: number[] = []
+
+			for (const { isCulled, maskedPageBounds } of editor.getRenderingShapes()) {
+				if (isCulled && maskedPageBounds) {
+					results.push(
+						// triangle 1
+						maskedPageBounds.minX,
+						maskedPageBounds.minY,
+						maskedPageBounds.minX,
+						maskedPageBounds.maxY,
+						maskedPageBounds.maxX,
+						maskedPageBounds.maxY,
+						// triangle 2
+						maskedPageBounds.minX,
+						maskedPageBounds.minY,
+						maskedPageBounds.maxX,
+						maskedPageBounds.minY,
+						maskedPageBounds.maxX,
+						maskedPageBounds.maxY
+					)
+				}
+			}
+
+			return results
+		})
+
+		return react('render culled shapes ', function renderCulledShapes() {
+			const canvas = canvasRef.current
+			if (!canvas) return
+
+			const width = canvas.clientWidth
+			const height = canvas.clientHeight
+			if (width !== canvas.width || height !== canvas.height) {
+				canvas.width = width
+				canvas.height = height
+				context.viewport(0, 0, width, height)
+			}
+
+			const verticesArray = shapeVertices.get()
+
+			context.clear(context.COLOR_BUFFER_BIT | context.DEPTH_BUFFER_BIT)
+
+			if (verticesArray.length > 0) {
+				const viewport = editor.getViewportPageBounds() // when the viewport changes...
+				context.uniform2f(viewportStartUniformLocation, viewport.minX, viewport.minY)
+				context.uniform2f(viewportEndUniformLocation, viewport.maxX, viewport.maxY)
+				const triangleGeoCpuBuffer = new Float32Array(verticesArray)
```

Comment:
@MitjaBezensek is this happening whenever the culled shapes change? we could probably pull this out, and re-use the same array, so we don't need to allocate so much memory each time

Reply:
Yes, this is happening all the time. I guess we would then need to be checking if the number of shapes changed and only create a new array in that case?

That said you can completely remove the culled shapes logic and it's hard to even notice any perf improvements since the whole thing is so fast.

Reply:
> I guess we would then need to be checking if the number of shapes changed and only create a new array in that case?

No, we can have the array be oversized, and always leave it the same size. Then in `drawArrays` we specify how many triangles there are to draw, so that it doesn't draw all of them

Reply:
Ah, I see 👍